packages/fxa-auth-server/lib/routes/mfa.ts

```diff
@@ -0,0 +1,339 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import * as isA from 'joi';
+import { Container } from 'typedi';
+import { StatsD } from 'hot-shots';
+import * as jwt from 'jsonwebtoken';
+import * as uuid from 'uuid';
+import { SecurityEvent } from 'fxa-shared/db/models/auth/security-event';
+import { Account } from 'fxa-shared/db/models/auth';
+import { AuthRequest, SessionTokenAuthCredential } from '../types';
+import { recordSecurityEvent } from './utils/security-event';
+import { ConfigType } from '../../config';
+import { OtpUtils } from './utils/otp';
+import * as AppError from '../error';
+
+/** Customs interface for mfa specific operations. */
+interface Customs {
+  check: (req: AuthRequest, email: string, action: string) => Promise<void>;
+  checkAuthenticated: (
+    req: AuthRequest,
+    uid: string,
+    email: string,
+    action: string
+  ) => Promise<void>;
+}
+
+/** Mailer interface for mfa specific operations  */
+interface Mailer {
+  sendVerifyAccountChangeEmail(
+    emails: string[],
+    account: Account,
+    emailOptions: {
+      uid: string;
+      code: string;
+      expirationTime: number;
+    }
+  ): Promise<void>;
+}
+
+/** DB interface for MFA specific db operations */
+interface DB {
+  account(uid): Promise<Account>;
+  securityEvent: (arg: SecurityEvent) => Promise<void>;
+}
+
+class MfaHandler {
+  constructor(
+    private readonly config: ConfigType,
+    private readonly db: DB,
+    private readonly otpUtils: OtpUtils,
+    private readonly customs: Customs,
+    private readonly mailer: Mailer,
+    private readonly statsd: StatsD
+  ) {}
+
+  async requestOtpCode(request: AuthRequest) {
+    const { uid, id: sessionTokenId } = request.auth
+      .credentials as SessionTokenAuthCredential;
+
+    const { action } = request.payload as { action: string };
+
+    const account = await this.db.account(uid);
+    await this.customs.checkAuthenticated(
+      request,
+      uid,
+      account.email,
+      'mfaOtpCodeRequest'
+    );
+
+    let success = false;
+    try {
+      const secret = `${account.uid}-${account.emailCode}-${sessionTokenId}-${action}`;
```

Comment:
This is different from the spec, `${account.uid}-${sessionToken.id}-${request.payload.action}`. Did you intend to include `emailCode`?

Reply:
Good call out. I decided to do this after looking more closely at how our other OTP codes that use the otplib function work. `emailCode` never leaves the server side, whereas `sessiontTokenId` and `uid` are both transmitted over the internet and potentially held in local cache. I think the fact that `emailCode` is never transmitted makes it a good thing to mix into the secret. I'll update the tech spec accordingly.